### ISSUE TITLE
Solution for "Or" and proof by cases 2.4.5 exercises (the last one)

### ISSUE THREAD

**Issue by lukaIvanic** [hrmacbeth/math2001#26]
<p>Let <span class="math notranslate nohighlight"><mjx-container class="MathJax CtxtMenu_Attached_0" jax="CHTML" tabindex="0" ctxtmenu_counter="370" style="font-size: 117.4%; position: relative;"><mjx-math class="MJX-TEX" aria-hidden="true"><mjx-mi class="mjx-i"><mjx-c class="mjx-c1D44E TEX-I"></mjx-c></mjx-mi></mjx-math><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>a</mi></math></mjx-assistive-mml></mjx-container></span> and <span class="math notranslate nohighlight"><mjx-container class="MathJax CtxtMenu_Attached_0" jax="CHTML" tabindex="0" ctxtmenu_counter="371" style="font-size: 117.4%; position: relative;"><mjx-math class="MJX-TEX" aria-hidden="true"><mjx-mi class="mjx-i"><mjx-c class="mjx-c1D44F TEX-I"></mjx-c></mjx-mi></mjx-math><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>b</mi></math></mjx-assistive-mml></mjx-container></span> be real numbers and suppose that <span class="math notranslate nohighlight"><mjx-container class="MathJax CtxtMenu_Attached_0" jax="CHTML" tabindex="0" ctxtmenu_counter="372" style="font-size: 117.4%; position: relative;"><mjx-math class="MJX-TEX" aria-hidden="true"><mjx-mi class="mjx-i"><mjx-c class="mjx-c1D44E TEX-I"></mjx-c></mjx-mi><mjx-mi class="mjx-i"><mjx-c class="mjx-c1D44F TEX-I"></mjx-c></mjx-mi><mjx-mo class="mjx-n" space="4"><mjx-c class="mjx-c3D"></mjx-c></mjx-mo><mjx-mi class="mjx-i" space="4"><mjx-c class="mjx-c1D44E TEX-I"></mjx-c></mjx-mi><mjx-mo class="mjx-n" space="4"><mjx-c class="mjx-c3D"></mjx-c></mjx-mo><mjx-mi class="mjx-i" space="4"><mjx-c class="mjx-c1D44F TEX-I"></mjx-c></mjx-mi></mjx-math><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>a</mi><mi>b</mi><mo>=</mo><mi>a</mi><mo>=</mo><mi>b</mi></math></mjx-assistive-mml></mjx-container></span>.  Show that either
<span class="math notranslate nohighlight"><mjx-container class="MathJax CtxtMenu_Attached_0" jax="CHTML" tabindex="0" ctxtmenu_counter="373" style="font-size: 117.4%; position: relative;"><mjx-math class="MJX-TEX" aria-hidden="true"><mjx-mi class="mjx-i"><mjx-c class="mjx-c1D44E TEX-I"></mjx-c></mjx-mi><mjx-mo class="mjx-n" space="4"><mjx-c class="mjx-c3D"></mjx-c></mjx-mo><mjx-mi class="mjx-i" space="4"><mjx-c class="mjx-c1D44F TEX-I"></mjx-c></mjx-mi><mjx-mo class="mjx-n" space="4"><mjx-c class="mjx-c3D"></mjx-c></mjx-mo><mjx-mn class="mjx-n" space="4"><mjx-c class="mjx-c30"></mjx-c></mjx-mn></mjx-math><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>a</mi><mo>=</mo><mi>b</mi><mo>=</mo><mn>0</mn></math></mjx-assistive-mml></mjx-container></span> or <span class="math notranslate nohighlight"><mjx-container class="MathJax CtxtMenu_Attached_0" jax="CHTML" tabindex="0" ctxtmenu_counter="374" style="font-size: 117.4%; position: relative;"><mjx-math class="MJX-TEX" aria-hidden="true"><mjx-mi class="mjx-i"><mjx-c class="mjx-c1D44E TEX-I"></mjx-c></mjx-mi><mjx-mo class="mjx-n" space="4"><mjx-c class="mjx-c3D"></mjx-c></mjx-mo><mjx-mi class="mjx-i" space="4"><mjx-c class="mjx-c1D44F TEX-I"></mjx-c></mjx-mi><mjx-mo class="mjx-n" space="4"><mjx-c class="mjx-c3D"></mjx-c></mjx-mo><mjx-mn class="mjx-n" space="4"><mjx-c class="mjx-c31"></mjx-c></mjx-mn></mjx-math><mjx-assistive-mml unselectable="on" display="inline"><math xmlns="http://www.w3.org/1998/Math/MathML"><mi>a</mi><mo>=</mo><mi>b</mi><mo>=</mo><mn>1</mn></math></mjx-assistive-mml></mjx-container></span>.</p>
<p>You will probably need the lemma <code class="docutils literal notranslate"><span class="pre">eq_zero_or_eq_zero_of_mul_eq_zero</span></code> proved in
<span class="reference internal" href="#solve-quadratic"><span class="std std-numref">Example 2.3.4</span>.</p>
<div class="highlight-lean notranslate"><div class="highlight"><pre><span></span><span class="kd">example</span> <span class="o">{</span><span class="n">a</span> <span class="n">b</span> <span class="o">:</span> <span class="n">ℝ</span><span class="o">}</span> <span class="o">(</span><span class="n">h1</span> <span class="o">:</span> <span class="n">a</span> <span class="bp">*</span> <span class="n">b</span> <span class="bp">=</span> <span class="n">a</span><span class="o">)</span> <span class="o">(</span><span class="n">h2</span> <span class="o">:</span> <span class="n">a</span> <span class="bp">*</span> <span class="n">b</span> <span class="bp">=</span> <span class="n">b</span><span class="o">)</span> <span class="o">:</span>
    <span class="n">a</span> <span class="bp">=</span> <span class="mi">0</span> <span class="bp">∧</span> <span class="n">b</span> <span class="bp">=</span> <span class="mi">0</span> <span class="bp">∨</span> <span class="n">a</span> <span class="bp">=</span> <span class="mi">1</span> <span class="bp">∧</span> <span class="n">b</span> <span class="bp">=</span> <span class="mi">1</span> <span class="o">:=</span> <span class="kd">by</span>
  <span class="gr">sorry</span>
</pre></div>
</div>
</li>

<br>

I’ve been really struggling with this exercise and would greatly appreciate it if someone could share their solution—or at least offer a hint.
Also, are solutions available somewhere? It would be super helpful to be able to compare my answers after finishing the exercises. I often feel my solutions end up a bit redundant, and my code tends to be "all over the place."

That said, I’ve been really enjoying the book—it’s exceptionally well-written!

**Comment by lukaIvanic** [comment]
Yeah thanks @rzeta0! I was mostly having syntax problems, but after revisiting the problem after a couple of days and with your hint, I got it in first try. Anyway, here's my messy code in case someone needs it, and in case someone has more concise code and is willing to share it, would be greatly appreciated! 


example {a b : ℝ} (h1 : a * b = a) (h2 : a * b = b):
    a = 0 ∧ b = 0 ∨ a = 1 ∧ b = 1 := by

    have h01 :=
      calc
        a * (b-1) = a * b - a := by ring
        _ = 0 := by addarith [h1]

    have h02 :=
      calc
        b * (a - 1) = a * b - b := by ring
        _ = 0 := by addarith [h2]

    have h3 := eq_zero_or_eq_zero_of_mul_eq_zero h01
    obtain ha | hb := h3
    left
    constructor
    apply ha
    calc
      b = a*b := by rw [h2]
      _ = 0 * b := by rw [ha]
      _ = 0 := by ring


    right
    constructor
    have hbb: b = 1 := by addarith [hb]
    have hab :=
      calc
        (a - 1) = (a-1) * (1/1) := by ring
        _ = (a-1) * (b/b) := by rw [hbb]
        _ =  (b * (a-1)) * (1/b) := by ring
        _ = (0) * (1/b) := by rw [h02]
        _ = 0 := by ring
    have haaa: a = 1 := by addarith [hab]
    apply haaa
    have b1 : b = 1 := by addarith [hb]
    apply b1


(I don't know why this is formatted weirdly)

**Comment by rzeta0** [comment]
good to see you found a solution.

by the way you may find it easier to discuss questions here: https://leanprover.zulipchat.com

it is a m much more active place with many experts (not me) willing to help
